AV JSON Importer and JSON key handler

ricdezen opened the pull request:
### Alpha Vantage Importer
Added an Importer for Alpha Vantage:
- Takes the datetime keys and inserts them in each atom
- Inserts the metadata in each atom
- Inserts the atom array in the database

### JSON key handler
Just a module containing a script that can apply recursively any kind of function to all the keys in a json dictionary, as deep as possible.
Also contains premade methods to rename keys and turn them into lower case.

Commit:
feat (importer) : added json_key_handler.py
+ json_key_handler.py : module that should contain static methods to manage keys in json objects, such as renaming them

Commit:
feat (json_key_handler.py): Generalized methods
+ Added generic method "apply_deep" taking in both a data dict/list and a function to apply to the keys
+ Added generic versions of apply_deep for dict and list
- Removed _lower_keys_dict and _lower_keys_list
~ lower_all_keys_deep now simply calls apply_deep with a lambda that lowers the parameter if its a string or leaves it untouched.

Commit:
feat(json_key_handler.py): added renaming method
+ Added method that recursively renames a bunch of keys, passed as a dictionary like:
{old_key : new_key ...}

CremaLuca committed:
doc(json_key_handler): added example documentation for rename_deep method

Commit:
fix (json_key_handler) : meh, I just hit ctrl+alt+L

Commit:
feat (av_json_data_importer.py): Added av importer class
+ Added class
+ Added renaming of the keys
+ The datetime key becomes a field of each atom
+ The metadata is added to each atom
+ The atoms are stripped from the dict as a list and inserted in the db

Commit:
fix (av_json_data_importer.py) : minor fix
~ Fixed a dictionary assignment that caused all "datetime" fields in atoms to have value "metadata" (lol)

feat (main.py) : added AV support
~ Moved around the directory selection
~ The script assumes the directory names are the keys in a constant dictionary at the top.

Commit:
fix (main.py) : added comment
+ Added comment stating the scripts behaviour regarding directories.

Commit:
fix : wrong table for insertion

CremaLuca commented on importer/av_json_data_importer.py:
```diff
@@ -0,0 +1,107 @@
+        atoms = self.__fix_atoms_datetime(atoms)
+        self.database.write(DatabaseData("atoms_b", atoms))
+
+    def _add_timestamp(self, data_dict: dict):
```
Is this method private? Doesn't private methods require two underscores?

https://www.geeksforgeeks.org/private-methods-in-python/

ricdezen replied:
My bad, not sure where but I somehow got the idea they needed only one, will fix it

CremaLuca commented on importer/av_json_data_importer.py:
```diff
@@ -0,0 +1,107 @@
+
+    def __add_metadata_to_atoms(self, metadata: dict, atoms: list):
+        '''
+        Adds each metadata field to all dictionary elements of the given list
```
A little confusing, try:
"Adds each metadata field to all ~dictionary~ atoms of the given list"

CremaLuca commented on importer/av_json_data_importer.py:
```diff
@@ -0,0 +1,107 @@
+
+    def _add_timestamp(self, data_dict: dict):
+        '''
+        Copies the timestamp key inside each atom.
```
```suggestion
        Copies the timestamp key inside each atom as a dictionary entry.
```

CremaLuca reviewed (changes requested):
Could you use `json_key_handler` on YahooJSONDataImporter too please?

ricdezen committed:
fix (av_json_importer.py) : spec fix

Co-Authored-By: Luca Crema <lc.crema@hotmail.com>

ricdezen commented:
> 
> 
> Could you use `json_key_handler` on YahooJSONDataImporter too please?

Omw
![vegitorunning](https://user-images.githubusercontent.com/56487105/79639653-54205b80-818d-11ea-8624-26ed7f3d3c57.gif)

Commit:
+ fix (json_key_handler.py): fixed missing underscores (private methods)
+ fix (av_json_data_importer.py): fixed missing underscores (private methods)
+ fix (yahoo_json_data_importer.py): now using json_key_handler

Commit:
Merge branch 'av_json_importer' of https://github.com/OTRI-Unipd/OTRI into av_json_importer

Commit:
fix (av_json_data_importer.py): minor spec fix

Commit:
fix (av_json_data_importer.py) : method call

CremaLuca reviewed (approved):
Seems all ok now. GJ
![OK](https://media1.giphy.com/media/kRXnZwKrPTwVq/200.webp?cid=ecf05e4728f2a60059608a8525c5f4f5e255d64737559ad6&rid=200.webp)
